core/dbt/task/init.py

```diff
@@ -84,29 +116,209 @@ def get_addendum(self, project_name, profiles_path):
+            "profile_template.yml and your supplied values. Run 'dbt debug' "
+            "to validate the connection."
+        )
+
+    def ask_for_adapter_choice(self) -> str:
+        """Ask the user which adapter (database) they'd like to use."""
+        available_adapters = list(_get_adapter_plugin_names())
+        prompt_msg = (
+            "Which database would you like to use?\n" +
+            "\n".join([f"[{n+1}] {v}" for n, v in enumerate(available_adapters)]) +
+            "\n\n(Don't see the one you want? https://docs.getdbt.com/docs/available-adapters)" +
+            "\n\nEnter a number"
+        )
+        numeric_choice = click.prompt(prompt_msg, type=click.INT)
+        return available_adapters[numeric_choice - 1]
+
+    def run(self):
+        """Entry point for the init task."""
+        profiles_dir = flags.PROFILES_DIR
         self.create_profiles_dir(profiles_dir)
-        if sample_adapter:
-            self.create_profiles_file(profiles_file, sample_adapter)
 
-        if os.path.exists(project_dir):
-            raise RuntimeError("directory {} already exists!".format(
-                project_dir
-            ))
+        try:
+            move_to_nearest_project_dir(self.args)
+            in_project = True
+        except dbt.exceptions.RuntimeException:
+            in_project = False
+
+        if in_project:
+            # When dbt init is run inside an existing project,
+            # just setup the user's profile.
+            logger.info("Setting up your profile.")
+            profile_name = self.get_profile_name_from_current_project()
+            profile_template_path = Path("profile_template.yml")
+            if profile_template_path.exists():
+                self.create_profile_using_profile_template()
+            else:
+                if not self.check_if_can_write_profile(profile_name=profile_name):
+                    return
+                adapter = self.ask_for_adapter_choice()
+                self.create_profile_from_scratch(
+                    adapter, profile_name=profile_name
+                )
+        else:
+            # When dbt init is run outside of an existing project,
+            # create a new project and set up the user's profile.
+            project_name = click.prompt("What is the desired project name?")
+            project_path = Path(project_name)
+            if project_path.exists():
+                logger.info(
+                    f"A project called {project_name} already exists here."
+                )
+                return
 
-        self.copy_starter_repo(project_dir)
+            self.copy_starter_repo(project_name)
+            os.chdir(project_name)
+            with open("dbt_project.yml", "r+") as f:
+                content = f"{f.read()}".format(
+                    project_name=project_name,
+                    profile_name=project_name
+                )
+                f.seek(0)
+                f.write(content)
+                f.truncate()
```

Comment:
>> Let's think about the "programmatic" entry-points to `init`, e.g. for the dbt Cloud IDE (which doesn't use the real `init` task today, but for consistency's sake, I'd feel better if it could). Do we need to preserve a "streamlined," CLI flag-based version of `init` that _just_ scaffolds the basic file structure, using a provided project name, and doesn't worry about `profiles.yml` at all?

> @jtcohen6 @leahwicz Just so I know where to head - should this PR need to concern itself with this, given that you've said dbt Cloud doesn't use the init task currently?

I think you've done it! This snippet of code is exactly what we'd want in the "programmatic" version. I could imagine wrapping this into a `create_project_files` method, re-adding a flag to init like `--starter-project-only`, and then supporting this workflow as:
```python
    def run(self):
        """Entry point for the init task."""
        if self.args.starter_project_only:
            project_name = self.args.starter_project_only
            create_project_files(project_name)
            return
            
        ... otherwise proceed interactively ...
```
```
dbt init --starter-project-only my_cool_project_name
```
Or, the equivalent interface into the dbt-core API ;)

@leahwicz What do you think? I don't think we need to make that change in this PR, but I'm satisfied knowing it would be a small lift.